File: 2644.py
# 촌수계산
# https://www.acmicpc.net/problem/2644

import sys

input = sys.stdin.readline

n = int(input())
s, e = map(int, input().split())
m = int(input())
graph = [[0] * (n + 1) for _ in range(n + 1)]

for _ in range(m):
    p, c = map(int, input().split())
    graph[p][c] = 1
    graph[c][p] = 1

queue = [(s, 0)]
visited = []
answer = -1
while len(queue) > 0:
    cur, count = queue.pop(0)

    for i in range(len(graph[cur])):
        if graph[cur][i] == 1:
            if i not in visited:
                if i == e:
                    answer = count + 1
                    break
                else:
                    visited.append(i)
                    queue.append((i, count + 1))

print(answer)

File: 2644.kt
// https://www.acmicpc.net/problem/2644
// 촌수 계산

import java.util.*

var start = 0
var dest = 0
var tree: MutableList<MutableList<Int>>? = null
var visited: MutableList<Int>? = null

fun main() {
    input()
    solve()
    println(visited?.get(dest))
}

fun input() {
    val reader = System.`in`.bufferedReader()
    val n = reader.readLine().toInt()
    tree = MutableList(n + 1) { mutableListOf() }
    visited = MutableList(n + 1) { -1 }
    val (a, b) = reader.readLine().split(" ").map { it.toInt() }
    start = a
    dest = b
    val m = reader.readLine().toInt()
    for (i in 0 until m) {
        val (aa, bb) = reader.readLine().split(" ").map { it.toInt() }
        tree?.let {
            it[aa].add(bb)
            it[bb].add(aa)
        }
    }
}

fun solve() {
    val queue = LinkedList<Int>()
    queue.add(start)
    visited?.let {
        it[start] = 0
    }
    loop1@ while (queue.isNotEmpty()) {
        val current = queue.poll()
        for (next in tree?.get(current) ?: return) {
            if (visited?.get(next) != -1) continue

            visited?.let { visited ->
                visited[next] = visited[current] + 1
            }

            if (next == dest) {
                break@loop1
            }

            queue.add(next)
        }
    }
}